diff --git a/drivers/platform/x86/apple-gmux.c b/drivers/platform/x86/apple-gmux.c
--- a/drivers/platform/x86/apple-gmux.c
+++ b/drivers/platform/x86/apple-gmux.c
@@ -958,6 +958,7 @@
 			pr_info("Switching to IGD");
 			gmux_switchto(VGA_SWITCHEROO_IGD);
 			vga_set_default_device(pdev);
+			pci_dev_put(pdev);
 		} else {
 			pr_err("force_idg is true, but couldn't find iGPU at 00:02.0! Is apple-set-os working?");
 		}
