diff --git a/include/linux/sched/bore.h b/include/linux/sched/bore.h
--- a/include/linux/sched/bore.h
+++ b/include/linux/sched/bore.h
@@ -45,5 +45,7 @@ extern int  sched_burst_protect_slice_lv_update_handler(const struct ctl_table *
 
 extern void reweight_entity(
 	struct cfs_rq *cfs_rq, struct sched_entity *se, unsigned long weight);
+extern void reweight_task_fair(struct rq *rq, struct task_struct *p,
+				const struct load_weight *lw);
 
 #endif /* _KERNEL_SCHED_BORE_H */
diff --git a/kernel/sched/bore.c b/kernel/sched/bore.c
--- a/kernel/sched/bore.c
+++ b/kernel/sched/bore.c
@@ -72,8 +72,12 @@ static void reweight_task_by_prio(struct task_struct *p, int prio) {
 	unsigned long weight = scale_load(sched_prio_to_weight[prio]);
 
 	if (se->on_rq) {
+		struct load_weight lw = {
+			.weight = weight,
+			.inv_weight = sched_prio_to_wmult[prio],
+		};
 		p->bore.stop_update = true;
-		reweight_entity(cfs_rq_of(se), se, weight);
+		reweight_task_fair(task_rq(p), p, &lw);
 		p->bore.stop_update = false;
 	} else
 		se->load.weight = weight;
diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -2063,8 +2063,8 @@ static inline bool do_preempt_weight(struct cfs_rq *cfs_rq,
 		return false;
 
 	if (static_branch_unlikely(&sched_burst_protect_slice_prefer_key)
-			? (pse->load.weight <= se->load.weight)
-			: (pse->load.weight <  se->load.weight))
+			? (pse->h_load.weight <= se->h_load.weight)
+			: (pse->h_load.weight <  se->h_load.weight))
 		return false;
 
 	if (!entity_eligible(cfs_rq, pse))
@@ -4869,8 +4869,8 @@ unsigned long __calc_prop_weight(struct cfs_rq *cfs_rq, struct sched_entity *se,
 	return max(weight, MIN_SHARES);
 }
 
-static void reweight_task_fair(struct rq *rq, struct task_struct *p,
-			       const struct load_weight *lw)
+void reweight_task_fair(struct rq *rq, struct task_struct *p,
+			const struct load_weight *lw)
 {
 	struct sched_entity *se = &p->se;
 	unsigned long weight = NICE_0_LOAD;
@@ -8289,7 +8289,7 @@ static bool dequeue_task_fair(struct rq *rq, struct task_struct *p, int flags)
 	struct cfs_rq *cfs_rq = cfs_rq_of(&p->se);
 	struct sched_entity *se = &p->se;
 	if ((flags & DEQUEUE_SLEEP) && entity_is_task(se)) {
-		if (cfs_rq->curr == se)
+		if (cfs_rq->h_curr == se)
 			update_curr(cfs_rq);
 		restart_burst_bore(p);
 	}
